gapic-generator/lib/gapic/presenters/service_presenter.rb

```diff
@@ -525,35 +545,59 @@ def mixins?
         @gem_presenter.mixins?
       end
 
-
       ##
-      # Whether this service presenter is a mixin inside a host service's gem
-      # (and not in its own gem)
+      # The models for the mixins
       #
-      # @return [Boolean]
+      # @return [Enumerable<::Gapic::Model::Mixins::Mixin>]
       #
-      def is_hosted_mixin?
-        Gapic::Model::Mixins.mixin_service_address? address, gem_address: @gem_presenter.address
+      def mixin_models
+        @gem_presenter.mixins_model.mixins
       end
+      
 
       ##
-      # Whether this service presenter is a mixin inside it's own gem
-      # (and not in another service's gem)
+      # Whether there are mixin services that this package has http binding overrides for.
       #
       # @return [Boolean]
       #
-      def is_main_mixin_service?
-        Gapic::Model::Mixins.mixin_service_address?(address) && !Gapic::Model::Mixins.mixin_service_address?(address, gem_address: @gem_presenter.address)
+      def mixin_binding_overrides?
+        mixin_presenters.any? { |mixin| !mixin.bindings_override.empty? }
       end
 
       ##
       # The mixin services that should be referenced
       # in the client for this service
       #
-      # @return [Enumerable<Gapic::Model::Mixins::Mixin>]
+      # @return [Enumerable<Gapic::Presenters::Service::MixinClientPresenter>]
       #
-      def mixins
-        @gem_presenter.mixins_model.mixins
+      def mixin_presenters
+        return [] unless mixins?
+        mixin_models.map do |model|
+          # we don't have mixin's descriptors loaded, so instead of starting with descriptor's methods
+          # and looking up the override, we'll start with overrides and see if any fit the mixin's namespace
+          bindings_override = begin
+            if @api.service_config&.http.nil?
+              {}
+            else
+              @api.service_config.http.rules.select do |http_rule|
+                 http_rule.selector.include? model.service
+              end.map do |http_rule|
```

Comment:
Just do

```ruby
end.to_h do |http_rule|
```

and remove the final to_h.

(or rubocop might catch this and tell you how to fix it.)

Reply:
:+1: